dfaker.py

```diff
@@ -0,0 +1,387 @@
+	"""
+	#solving blood glucose eqn 
+	glucose = solution[:, 1]
+	time = solution[:, 2]
+	#smoothing blood glucose eqn
+	lowess = sm.nonparametric.lowess
+	smoothing_distance = 1.3 #1.3 minutes
+	fraction = (smoothing_distance / (params['num_days'] * 60 * 24)) * 100
+	result = lowess(glucose, time, frac=fraction, is_sorted=True)
+	
+	if params['gaps']:
+		result = make_gaps(params, result)
+	smoothed_time = result[:, 0]
+	smoothed_glucose = result[:, 1]
+	return smoothed_glucose, smoothed_time
+
+def get_offset(params):
+	utc_tz = timezone('UTC')
+	local_tz = timezone(params['zone'])
+	naive = datetime(params['year'], params['month'], params['day'], params['hour'], params['minute'])
+	offset = int((utc_tz.localize(naive) - local_tz.localize(naive)) / timedelta(minutes=1))
+	return offset
+
+def make_gaps(params, time_gluc):
+	gaps = random.randint(4 * params['num_days'], 5 * params['num_days']) # amount of gaps	
+	for _ in range(gaps):
+		gap_length = random.randint(6, 36) # length of gaps in 5-min segments
+		start_index = random.randint(0, len(time_gluc))	
+		if start_index + gap_length > len(time_gluc):
+			end_index = len(time_gluc) - 5
+		else:
+			end_index = start_index + gap_length
+		time_gluc = np.delete(time_gluc, time_gluc[start_index:end_index][::2], 0)
+	return time_gluc
+
+def generate_boluses(solution, start_time):
+	""" Generates events for both bolus enteries and wizard enteries.
+		Returns carb, time and glucose values for each event
+	"""
+	all_carbs = solution[:, 0]
+	glucose = solution[:,1]
+	time = solution[:, 2]
+	ts = make_timesteps(start_time, solution[:,2])
+	positives = []
+	for row in zip(all_carbs, ts, glucose):
+		if row[0] > 10:
+			positives.append(row)
+	np_pos = np.array(clean_up_boluses(positives))
+	cleaned = remove_night_boluses(np_pos)
+	for row in cleaned:
+		if row[0] > 120:
+			row[0] = row[0] / 2	
+		elif row[0] > 30:
+			row[0] = row[0] * 0.75
+	bolus, wizard = bolus_or_wizard(cleaned)
+	np_bolus, np_wizard = np.array(bolus), np.array(wizard)
+	b_carbs, b_ts  = np_bolus[:, 0], np_bolus[:, 1]
+	w_card, w_ts, w_gluc = np_wizard[:, 0], np_wizard[:, 1], np_wizard[:,2]
+	return b_carbs, b_ts, w_card, w_ts, w_gluc
+
+def clean_up_boluses(carb_time, filter_rate=7):
+	""" Cleans up clusters of carb events based on meal freqeuncy
+		carb_time -- a numpy array of carbohydrates at different points in time
+		filter_rate -- how many 5 minute segments should be filtered within the range
+					   of a single event.
+					   Example: 7 * 5 --> remove the next 35 minutes of data from carb_time
+	"""
+	return carb_time[::filter_rate]
+
+def remove_night_boluses(carb_time_gluc):
+	"""Removes night boluses excpet for events with high glucose""" 
+	keep = []
+	for row in carb_time_gluc:
+		hour = time.gmtime(row[1]).tm_hour
+		if (hour > 6 and hour < 23):
+			keep.append(row)
+		elif int(row[2]) not in range(0,250):
+			keep.append(row)
+	np_keep = np.array(keep)
+	return np_keep
+
+def bolus_or_wizard(solution):
```

Comment:
What's the purpose here? I'm guessing to decide when to generate wizards linked with boluses and when to just have plain boluses, but I'm not sure...